### PR TITLE
Add mosaic data augmentation

### LINKED ISSUE #250: Mosaic Augmentation for Object Detection
Mosaic augmentation for object detection is used in Yolo-V4 literature, [FR](https://github.com/keras-team/keras-cv/issues/21). I'm not sure if it's used for classification tasks there. I was wondering if it's possible to do so for the image classification task.  Here are two issues:

 - **References. Not sure if it's ever used for classification only in any literature.**
 - **Creation of class labels.** 

For creating class labels, here is one possible solution, described [HERE](https://stackoverflow.com/a/65182855/9215780); using Dirichlet distribution. 

```python
# for 2 images. Equivalent to λ and (1-λ)
>>> np.random.dirichlet((1, 1), 1)  
array([[0.92870347, 0.07129653]])  

>>> np.random.dirichlet((1, 1, 1), 1)  # for 3 images.
array([[0.38712673, 0.46132787, 0.1515454 ]])

>>> np.random.dirichlet((1, 1, 1, 1), 1)  # for 4 images.
array([[0.59482542, 0.0185333 , 0.33322484, 0.05341645]])
```

As mosaic takes 4 images. 
inbox_1984321_b0530886377eb3d7f0c29401ba069ffa_download (1)

---

update:

Mosaic augmentation for classification did use in literature. Please check: https://github.com/keras-team/keras-cv/issues/250#issuecomment-1130256698

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?

Introduces the Mosaic Augmentation

Fixes #250 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@LukeWood 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: keras_cv/layers/preprocessing/mosaic.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    def __init__(
+        self, offset=(0.25, 0.75), bounding_box_format=None, seed=None, **kwargs
+    ):
+        super().__init__(seed=seed, **kwargs)
+        self.offset = offset
+        self.bounding_box_format = bounding_box_format
+        self.center_sampler = preprocessing.parse_factor(offset)
+        self.seed = seed
+
+    def _batch_augment(self, inputs):
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/mosaic.py:77]: when I attempt to train a RetinaNet with this I get NaNs- which indicates that some of the bounding boxes that are fully out of the image are still being included in the dataset.

Lets implement a utility in `keras_cv/bounding_box/utils.py` to filter bounding boxes out of a given image

**Review by LukeWood — CHANGES_REQUESTED** [review]
VIsually - looks great, functionally: issue with NaNs

File: keras_cv/layers/preprocessing/mosaic.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        seed: Integer. Used to create a random seed.
+
+    References:
+        - [Yolov4 paper](https://arxiv.org/pdf/2004.10934).
+        - [Yolov5 implementation](https://github.com/ultralytics/yolov5).
+        - [YoloX implementation](https://github.com/Megvii-BaseDetection/YOLOX)
+
+    Sample usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/mosaic.py:59]: we can also use keras_cv.datasets to show how to do this for pascalVOC

File: keras_cv/layers/preprocessing/mosaic.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            axis=-1,
+        )
+        translate_values = tf.expand_dims(translate_values, axis=1)
+
+        # translating boxes
+        boxes_for_mosaic = boxes_for_mosaic + translate_values
+        boxes_for_mosaic = tf.concat([boxes_for_mosaic, rest], axis=-1)
+        boxes_for_mosaic = tf.reshape(boxes_for_mosaic, [-1, bounding_boxes.shape[-1]])
+
+        boxes_for_mosaic = bounding_box.clip_to_image(
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/mosaic.py:250]: basically bounding boxes that are shifted out of bounds will need to have their class_id set to -1.

we should update clip_to_image() to also filter out boxes that are fully outside of the image - and assign them -1 class ids.

File: keras_cv/layers/preprocessing/mosaic.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/mosaic.py:77]: Actually, I pushed a fix

File: keras_cv/layers/preprocessing/mosaic.py (hunk 2 above)
Comment (quantumalaviya) [reply keras_cv/layers/preprocessing/mosaic.py:59]: That would make sense. Do you think updating the examples/layers/preprocessing/bounding_box/demo_utils.py would be a better idea here?

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
Thank you @quantumalaviya for all of your hard work in getting `layers.Mosaic()` done.

super powerful augmentation; excited to have it in the package!

**Review by LukeWood — CHANGES_REQUESTED** [review]
Getting NaNs in training

**Comment by LukeWood** [comment]
I have isolated the NaN issue to be related to RetinaNet, not Mosaic.  

/gcbrun

**Comment by LukeWood** [comment]
https://github.com/keras-team/keras-cv/issues/845

**Comment by LukeWood** [comment]
Thanks @quantumalaviya for your persistence in contributing this augmentation!

Apologies for the confusion as to the bug source!
